test/method.jl

```diff
@@ -0,0 +1,181 @@
+        # but they have different method objects. And different (but equivelent) ASTd
+        # this generates something that should be the same as what `signature(f16)` does
+        # but with a gensym'd variable name
+        f16_alt(x::Array{<:Real, 1}) = 2x
+        f16_alt_sig = signature(first(methods(f16_alt)))
+        @test f16_alt_sig[:name] == :f16_alt
+        @test occursin(  # Hack to deal with gensymed name. Make it a string and use regex
+            r"^Expr\[:\(x::\(Array\{var\"(.*?)\", 1\} where var\"\1\" <: Real\)\)\]$",
+            string(f16_alt_sig[:args])
+        )
+    end
+
+    @testset "anon functions" begin
+        @test_signature (x) -> x  # no args
+        @test_signature (x)->2x
+
+        @test_signature ((::T) where T) -> 0   # Anonymous parameter
+
+
+    end
+
+    @testset "vararg" begin
+        @test_signature f17(xs::Vararg{Any, N} where N) = 2
+
+        # `f17_alt(xs...) = 2` lowers to the same method as `f18`
+        # but has a different AST according to `splitdef` so we can't us @test_signature
+        f17_alt(xs...) = 2
+        test_matches(
+            signature(only_method(f17_alt)),
+            Dict(
+                :name => :f17_alt,
+                :args => [:(xs::(Vararg{Any, N} where N))]
+            )
+        )
+
+        @test_signature f18(xs::Vararg{Int64, N} where N) = 2
+        @test_signature f19(x, xs::Vararg{Any, N} where N) = 2x
+    end
+
+    @testset "kwargs" begin  # We do not support them right now
+        #Following is broken:
+        #@test_signature kwargs17(x; y=3x) = 2x
+        kwargs17(x; y=3x) = 2x
+        # at least be sure we get the rest right:
+        test_matches(
+            signature(only_method(kwargs17)),
+            Dict(
+                :name => :kwargs17,
+                :args => [:x]
+            )
+        )
+    end
+
+
+    @testset "Constructors (basic)" begin
+        # demo type for testing on
+        struct NoParamStruct
+            x
+        end
+
+        test_matches(  # default constructor
+            signature(only_method(NoParamStruct, Tuple{Any})),
+            Dict(:name => :NoParamStruct, :args => [:x]),
+        )
+
+
+        @test_signature(
+            NoParamStruct(x::Bool, y::Int32) = NoParamStruct(x || y > 2),
+            only_method(NoParamStruct, Tuple{Bool, Int32})
+        )
+
+        struct OneParamStructBasic{T}
+            x::T
+        end
+
+        test_matches(  # default constructor
+            signature(only_method(OneParamStructBasic, Tuple{Any})),
+            Dict(:name => :OneParamStructBasic, :args => [:(x::T)]),
+        )
+
+        @test_signature(
+            OneParamStructBasic(x::Bool, y::Int32) = OneParamStruct(x || y > 2),
+            only_method(OneParamStructBasic, Tuple{Bool, Int32})
+        )
+    end
+
+    @testset "params (via Constructors with type params" begin
+        struct OneParamStruct{T}
+            x::T
+        end
```

Comment:
I'd be fine if you move the `OneParamStructBasic` into this testset

Reply:
I don't see any reason to?

Reply:
It would avoid having to create an extra struct

Reply:
I think I prefer the seperation. structs are cheap

Reply:
👌